stellarphot/io/aavso.py

```diff
@@ -0,0 +1,253 @@
+            `AAVSOExtendedFileFormatColumns`.
+        star_id : str, optional
+            ID of the star to be written to the file. If not provided, all stars
+            in the table will be written to the file.
+        """
+        if star_id is not None:
+            use_data = data[data['id'] == star_id]
+        else:
+            use_data = data
+        for source_column, destination_column in column_map.items():
+            if (self.ensemble and
+                (destination_column == AAVSOExtendedFileFormatColumns.COMP_STAR_MAG or
+                 destination_column == AAVSOExtendedFileFormatColumns.COMP_STAR_NAME)):
+                raise ValueError("Cannot set comparison star magnitude or name for ensemble observation")
+
+            try:
+                setattr(self, destination_column.value, use_data[source_column])
+            except AttributeError:
+                raise AttributeError(f"Column {destination_column.value} not allowed in AAVSO extended format")
+            except KeyError:
+                raise KeyError(f"Column {source_column} not found in data")
+
+    def set_column_from_single_value(self, value, column):
+        setattr(self, column, Column([value] * len(self.variable_mag), name=column))
+
+    def to_table(self):
+        """
+        Convert the file to an astropy table.
+
+        Returns
+        -------
+        table : `astropy.table.Table`
+            The table representation of the file. Note that there is important information
+            in the table's meta attribute.
+        """
+        # Make a table
+        table_description = resources.read_text('stellarphot.io', 'aavso_submission_schema.yml')
+        table_structure = yaml.safe_load(table_description)
+
+        table_dict = {}
+        length = 0
+
+        # Begin setting up table structure, with column names as keys
+        # and the column data as values. If the "column" is really
+        # a single item, then make a single entry as the value.
+        # It will be turned into a column eventually.
+        for key in table_structure['data'].keys():
+            # woot! my first walrus operator!
+            if isinstance((item := getattr(self, key.lower())), Column):
+                # We have a column, but it might not have been set.
+                if len(item) == 0:
+                    item = "na"
+                table_dict[key] = item
+                # Kepp track of the longest column
+                length = len(item) if len(item) > length else length
+            else:
+                table_dict[key] = item if len(item) > 0 else "na"
+
+        # Fix up anything that is supposed to be a column and isn't yet.
+        for k, v in table_dict.items():
+            if not isinstance(v, Column) or len(v) != length:
+                # If the length is not correct assume we have a single value
+                # and make a column out of it.
+                table_dict[k] = Column([v] * length, name=k)
```

Comment:
But I assume this 'Fix up anything' block has to be separate from the previous block of code because you need the longest `length` determined first.

However, it is unclear to me why you expect varying lengths for your input data. If I understand it correctly, the previous block of code (where `length` gets set) is checking the attributes matching the lowercase version of the column names listed in the schema file, and checking if they are astropy columns.  If they are, it checks their length and it is longer than the longest known previous column, sets `length to that value, finally adding them to `table_dict`.  If not a Column, it adds them directly to `table_dict`. Those non-columns get turned into columns by turning them into lists with the value  repeating `length` times.  But those original Columns apparently do not have their lengths adjusted.

In any case, how can you have columns with different lengths in the same astropy Table?

Reply:
> But I assume this 'Fix up anything' block has to be separate from the previous block of code because you need the longest length determined first.

Yep, exactly.

> If not a Column, it adds them directly to table_dict. Those non-columns get turned into columns by turning them into lists with the value  repeating length` times. But those original Columns apparently do not have their lengths adjusted.

Yes -- at the end of this block the dict values are either columns whose length we haven't checked or Column of a signle value of the correct length.

> In any case, how can you have columns with different lengths in the same astropy Table?

You can't which is why it is a dict up to this point. 

Calling the function right now is not trivial, and looks something [like this](https://github.com/feder-observatory/stellarphot/pull/146/files#diff-48c06c5d761d53093f4f352219670fc2689a171205bbf5d9617d4dc4cc60b306R27)...you create the object, then set a few attributes, then set the properties for the variable star magnitude (including a slice of the photometry data), then set properties for the check star, etc.

My fear was that some slicing might go awry, hence the assumption that lengths might not be the same.

Once this is in and released on 1.x, I'll also port it to 2.0, and use the new `PhotometryData` structure as the input. That will mean a substantial rewrite of this but it should be simpler since the `PhotometryData` is nicely structured.